### PR TITLE
Chore(EC2-checks): Update the ec2 terminate with tags experiment to target only running instances

### PR DESCRIPTION
**Pull request by uditgaurav**
Signed-off-by: uditgaurav <udit@chaosnative.com>

####  This PR is for:

- Update the pre and post chaos check for ec2-terminate using tags.
   - `pre-chaos`: Will check if any running instance found filtered from given instance tag. If no instance found it will fail. So it will now manage the case where the filtered set of instances are both running and non-running instances.
   - `post-chaos`: Status check of the target instance under chaos (IUC). It will fail if the target instance is not in running state post chaos.

### CODE REVIEW

**Comment by uditgaurav** [comment]
#### Experiment Logs:

```bash
udit@ubuntu ~/g/s/g/l/litmus-go> kubectl logs -f ec2-terminate-by-tag-bg05mm-87djn
W0503 22:09:40.528219       1 client_config.go:541] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
time="2021-05-03T22:09:40Z" level=info msg="Experiment Name: ec2-terminate-by-tag"
time="2021-05-03T22:09:40Z" level=info msg="[PreReq]: Getting the ENV for the  experiment"
time="2021-05-03T22:09:40Z" level=info msg="[PreReq]: Updating the chaos result of ec2-terminate-by-tag experiment (SOT)"
time="2021-05-03T22:09:40Z" level=info msg="The instance information is as follows" Instance Tag="team:devops" Instance Affected Percentage=100 Sequence=parallel Chaos Duration=30 Chaos Namespace=default Ramp Time=0
time="2021-05-03T22:09:40Z" level=info msg="[Status]: Verify that the AUT (Application Under Test) is running (pre-chaos)"
time="2021-05-03T22:09:40Z" level=info msg="[Status]: No appLabels provided, skipping the application status checks"
time="2021-05-03T22:09:41Z" level=info msg="[Info]: Targeting the running instances filtered from instance tag" Number of running instances filtered=2 Total number of instances filtered=3
time="2021-05-03T22:09:41Z" level=info msg="[Chaos]:Number of Instance targeted: 2"
time="2021-05-03T22:09:41Z" level=info msg="[Info]: Target instanceID list, [i-0af1943d2154b8d43 i-0766e87263d8b6aa6]"
time="2021-05-03T22:09:41Z" level=info msg="[Chaos]: Stoping the desired EC2 instance"
time="2021-05-03T22:09:41Z" level=info msg="Stopping ec2 instance:" PreviousState=running InstanceId=i-0af1943d2154b8d43 CurrentState=stopping
time="2021-05-03T22:09:41Z" level=info msg="[Chaos]: Stoping the desired EC2 instance"
time="2021-05-03T22:09:42Z" level=info msg="Stopping ec2 instance:" PreviousState=running InstanceId=i-0766e87263d8b6aa6 CurrentState=stopping
time="2021-05-03T22:09:42Z" level=info msg="[Wait]: Wait for EC2 instance 'i-0af1943d2154b8d43' to come in stopped state"
time="2021-05-03T22:09:42Z" level=info msg="[Status]: Checking EC2 instance status"
time="2021-05-03T22:09:42Z" level=info msg="The instance state is stopping"
time="2021-05-03T22:09:44Z" level=info msg="The instance state is stopping"
time="2021-05-03T22:09:46Z" level=info msg="The instance state is stopping"
time="2021-05-03T22:09:48Z" level=info msg="The instance state is stopping"
time="2021-05-03T22:09:50Z" level=info msg="The instance state is stopping"
time="2021-05-03T22:09:53Z" level=info msg="The instance state is stopping"
time="2021-05-03T22:09:55Z" level=info msg="The instance state is stopping"
time="2021-05-03T22:09:57Z" level=info msg="The instance state is stopping"
time="2021-05-03T22:09:59Z" level=info msg="The instance state is stopping"
time="2021-05-03T22:10:01Z" level=info msg="The instance state is stopping"
time="2021-05-03T22:10:03Z" level=info msg="The instance state is stopping"
time="2021-05-03T22:10:05Z" level=info msg="The instance state is stopped"
time="2021-05-03T22:10:07Z" level=info msg="[Wait]: Wait for EC2 instance 'i-0766e87263d8b6aa6' to come in stopped state"
time="2021-05-03T22:10:07Z" level=info msg="[Status]: Checking EC2 instance status"
time="2021-05-03T22:10:08Z" level=info msg="The instance state is stopping"
time="2021-05-03T22:10:10Z" level=info msg="The instance state is stopping"
time="2021-05-03T22:10:12Z" level=info msg="The instance state is stopped"
time="2021-05-03T22:10:14Z" level=info msg="[Wait]: Waiting for chaos interval of 30s"
time="2021-05-03T22:10:44Z" level=info msg="[Chaos]: Starting back the EC2 instance"
time="2021-05-03T22:10:44Z" level=info msg="Starting ec2 instance:" CurrentState=pending PreviousState=stopped InstanceId=i-0af1943d2154b8d43
time="2021-05-03T22:10:44Z" level=info msg="[Chaos]: Starting back the EC2 instance"
time="2021-05-03T22:10:45Z" level=info msg="Starting ec2 instance:" CurrentState=pending PreviousState=stopped InstanceId=i-0766e87263d8b6aa6
time="2021-05-03T22:10:45Z" level=info msg="[Wait]: Wait for EC2 instance 'i-0af1943d2154b8d43' to get in running state"
time="2021-05-03T22:10:45Z" level=info msg="[Status]: Checking EC2 instance status"
time="2021-05-03T22:10:45Z" level=info msg="The instance state is pending"
time="2021-05-03T22:10:47Z" level=info msg="The instance state is pending"
time="2021-05-03T22:10:49Z" level=info msg="The instance state is pending"
time="2021-05-03T22:10:52Z" level=info msg="The instance state is pending"
time="2021-05-03T22:10:54Z" level=info msg="The instance state is pending"
time="2021-05-03T22:10:56Z" level=info msg="The instance state is pending"
time="2021-05-03T22:10:58Z" level=info msg="The instance state is running"
time="2021-05-03T22:11:00Z" level=info msg="[Wait]: Wait for EC2 instance 'i-0766e87263d8b6aa6' to get in running state"
time="2021-05-03T22:11:00Z" level=info msg="[Status]: Checking EC2 instance status"
time="2021-05-03T22:11:00Z" level=info msg="The instance state is running"
time="2021-05-03T22:11:02Z" level=info msg="[Chaos]: Time is up for experiment: ec2-terminate-by-tag"
time="2021-05-03T22:11:02Z" level=info msg="[Confirmation]: EC2 terminate chaos has been injected successfully"
time="2021-05-03T22:11:03Z" level=info msg="[Status]: EC2 instance is in running state (post chaos)"
time="2021-05-03T22:11:03Z" level=info msg="[Status]: Verify that the AUT (Application Under Test) is running (post-chaos)"
time="2021-05-03T22:11:03Z" level=info msg="[Status]: No appLabels provided, skipping the application status checks"
time="2021-05-03T22:11:03Z" level=info msg="[The End]: Updating the chaos result of ec2-terminate-by-tag experiment (EOT)"
```

**Comment by uditgaurav** [comment]
#### Experiment Logs:

```bash
udit@ubuntu ~/g/s/g/l/litmus-go> kubectl logs -f ec2-terminate-by-tag-yg03cj-dfzbc
W0503 22:32:11.754813       1 client_config.go:541] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
time="2021-05-03T22:32:11Z" level=info msg="Experiment Name: ec2-terminate-by-tag"
time="2021-05-03T22:32:11Z" level=info msg="[PreReq]: Getting the ENV for the  experiment"
time="2021-05-03T22:32:11Z" level=info msg="[PreReq]: Updating the chaos result of ec2-terminate-by-tag experiment (SOT)"
time="2021-05-03T22:32:11Z" level=info msg="The instance information is as follows" Chaos Namespace=default Ramp Time=0 Instance Tag="team:devops" Instance Affected Percentage=100 Sequence=parallel Chaos Duration=30
time="2021-05-03T22:32:11Z" level=info msg="[Status]: Verify that the AUT (Application Under Test) is running (pre-chaos)"
time="2021-05-03T22:32:11Z" level=info msg="[Status]: No appLabels provided, skipping the application status checks"
time="2021-05-03T22:32:12Z" level=info msg="[Info]: Targeting the running instances filtered from instance tag" Total number of instances filtered=3 Number of running instances filtered=2
time="2021-05-03T22:32:12Z" level=info msg="[Chaos]:Number of Instance targeted: 2"
time="2021-05-03T22:32:12Z" level=info msg="[Info]: Target instanceID list, [i-0766e87263d8b6aa6 i-0af1943d2154b8d43]"
time="2021-05-03T22:32:12Z" level=info msg="[Chaos]: Stopping the desired EC2 instance"
time="2021-05-03T22:32:13Z" level=info msg="Stopping ec2 instance:" CurrentState=stopping PreviousState=running InstanceId=i-0766e87263d8b6aa6
time="2021-05-03T22:32:13Z" level=info msg="[Chaos]: Stopping the desired EC2 instance"
time="2021-05-03T22:32:13Z" level=info msg="Stopping ec2 instance:" CurrentState=stopping PreviousState=running InstanceId=i-0af1943d2154b8d43
time="2021-05-03T22:32:13Z" level=info msg="[Wait]: Wait for EC2 instance 'i-0766e87263d8b6aa6' to get in stopped state"
time="2021-05-03T22:32:13Z" level=info msg="[Status]: Checking EC2 instance status"
time="2021-05-03T22:32:14Z" level=info msg="The instance state is stopping"
time="2021-05-03T22:32:16Z" level=info msg="The instance state is stopping"
time="2021-05-03T22:32:18Z" level=info msg="The instance state is stopping"
time="2021-05-03T22:32:20Z" level=info msg="The instance state is stopping"
time="2021-05-03T22:32:22Z" level=info msg="The instance state is stopping"
time="2021-05-03T22:32:24Z" level=info msg="The instance state is stopping"
time="2021-05-03T22:32:27Z" level=info msg="The instance state is stopping"
time="2021-05-03T22:32:29Z" level=info msg="The instance state is stopping"
time="2021-05-03T22:32:31Z" level=info msg="The instance state is stopping"
time="2021-05-03T22:32:33Z" level=info msg="The instance state is stopping"
time="2021-05-03T22:32:35Z" level=info msg="The instance state is stopping"
time="2021-05-03T22:32:37Z" level=info msg="The instance state is stopped"
time="2021-05-03T22:32:39Z" level=info msg="[Wait]: Wait for EC2 instance 'i-0af1943d2154b8d43' to get in stopped state"
time="2021-05-03T22:32:39Z" level=info msg="[Status]: Checking EC2 instance status"
time="2021-05-03T22:32:40Z" level=info msg="The instance state is stopping"
time="2021-05-03T22:32:42Z" level=info msg="The instance state is stopped"
time="2021-05-03T22:32:44Z" level=info msg="[Wait]: Waiting for chaos interval of 30s"
time="2021-05-03T22:33:14Z" level=info msg="[Chaos]: Starting back the EC2 instance"
time="2021-05-03T22:33:14Z" level=info msg="Starting ec2 instance:" CurrentState=pending PreviousState=stopped InstanceId=i-0766e87263d8b6aa6
time="2021-05-03T22:33:14Z" level=info msg="[Chaos]: Starting back the EC2 instance"
time="2021-05-03T22:33:15Z" level=info msg="Starting ec2 instance:" InstanceId=i-0af1943d2154b8d43 CurrentState=pending PreviousState=stopped
time="2021-05-03T22:33:15Z" level=info msg="[Wait]: Wait for EC2 instance 'i-0766e87263d8b6aa6' to get in running state"
time="2021-05-03T22:33:15Z" level=info msg="[Status]: Checking EC2 instance status"
time="2021-05-03T22:33:15Z" level=info msg="The instance state is pending"
time="2021-05-03T22:33:17Z" level=info msg="The instance state is pending"
time="2021-05-03T22:33:19Z" level=info msg="The instance state is pending"
time="2021-05-03T22:33:22Z" level=info msg="The instance state is pending"
time="2021-05-03T22:33:24Z" level=info msg="The instance state is pending"
time="2021-05-03T22:33:26Z" level=info msg="The instance state is pending"
time="2021-05-03T22:33:28Z" level=info msg="The instance state is pending"
time="2021-05-03T22:33:30Z" level=info msg="The instance state is running"
time="2021-05-03T22:33:32Z" level=info msg="[Wait]: Wait for EC2 instance 'i-0af1943d2154b8d43' to get in running state"
time="2021-05-03T22:33:32Z" level=info msg="[Status]: Checking EC2 instance status"
time="2021-05-03T22:33:32Z" level=info msg="The instance state is running"
time="2021-05-03T22:33:34Z" level=info msg="[Chaos]: Time is up for experiment: ec2-terminate-by-tag"
time="2021-05-03T22:33:34Z" level=info msg="[Confirmation]: EC2 terminate chaos has been injected successfully"
time="2021-05-03T22:33:35Z" level=info msg="[Status]: EC2 instance is in running state (post chaos)"
time="2021-05-03T22:33:35Z" level=info msg="[Status]: Verify that the AUT (Application Under Test) is running (post-chaos)"
time="2021-05-03T22:33:35Z" level=info msg="[Status]: No appLabels provided, skipping the application status checks"
time="2021-05-03T22:33:35Z" level=info msg="[The End]: Updating the chaos result of ec2-terminate-by-tag experiment (EOT)"
```

File: chaoslib/litmus/ec2-terminate-by-tag/lib/ec2-terminate-by-tag.go [hunk 1]
```diff
@@ -230,22 +224,49 @@ func CalculateInstanceAffPerc(InstanceAffPerc int, instanceList []string) []stri
...
+			return err
+		}
+		if instanceState != "running" {
+			return errors.Errorf("failed to get the ec2 instance '%v' in running sate, current state: %v", id, instanceState)
+		}
+	}
+	return nil
+}
+
+//SetTargetInstance will select the target instance which are in running sate and filtered from the given instance tag
```
Comment (ispeakc0de) [inline chaoslib/litmus/ec2-terminate-by-tag/lib/ec2-terminate-by-tag.go:242]:
```suggestion
//SetTargetInstance will select the target instance which are in running state and filtered from the given instance tag
```

**Review by ksatchit — APPROVED** [review]

File: experiments/kube-aws/ec2-terminate-by-tag/experiment/ec2-terminate-tag.go [hunk 2]
```diff
@@ -126,15 +126,14 @@ func EC2TerminateByTag(clients clients.ClientSets) {
 		events.GenerateEvents(&eventsDetails, clients, &chaosDetails, "ChaosEngine")
 	}
 
-	//Verify the aws ec2 instance is running (pre chaos)
-	err = litmusLIB.InstanceStatusCheckByTag(experimentsDetails.InstanceTag, experimentsDetails.Region)
+	//selecting the target instance (pre chaos)
+	err = litmusLIB.SetTargetInstance(&experimentsDetails)
```
Comment (ispeakc0de) [inline experiments/kube-aws/ec2-terminate-by-tag/experiment/ec2-terminate-tag.go:130]: Can we move this function to chaoslib?

File: chaoslib/litmus/ec2-terminate-by-tag/lib/ec2-terminate-by-tag.go [hunk 3]
```diff
@@ -230,22 +224,49 @@ func CalculateInstanceAffPerc(InstanceAffPerc int, instanceList []string) []stri
 	return newIDList
 }
 
-//InstanceStatusCheckByTag is used to check the instance status of all the instance under chaos.
-func InstanceStatusCheckByTag(instanceTag, region string) error {
+//PostChaosInstanceStatusCheckByTag is used to check the instance status of the instances under chaos post chaos.
+func PostChaosInstanceStatusCheckByTag(targetInstanceIDList []string, region string) error {
```
Comment (ispeakc0de) [inline chaoslib/litmus/ec2-terminate-by-tag/lib/ec2-terminate-by-tag.go:228]: Can we rename the func name? it is no longer selecting instance via tag inside this function instead it received the instance list

File: experiments/kube-aws/ec2-terminate-by-tag/experiment/ec2-terminate-tag.go (hunk 2 above)
Comment (ksatchit) [reply experiments/kube-aws/ec2-terminate-by-tag/experiment/ec2-terminate-tag.go:130]: more of a pkg/aws candidate, would prefer moving that separately.

**Review by ksatchit — APPROVED** [review]

**Review by ispeakc0de — APPROVED** [review]

**Pull request merged**